File: src/main/java/com/matag/game/turn/phases/PhaseFactory.java
package com.matag.game.turn.phases;

import java.util.List;

import org.springframework.stereotype.Component;

import lombok.AllArgsConstructor;

@AllArgsConstructor
@Component
public class PhaseFactory {
  private final List<Phase> phases;

  public Phase get(String phaseName) {
    return phases.stream()
      .filter(phase -> phase.getName().equals(phaseName))
      .findFirst()
      .orElseThrow(() -> new UnsupportedOperationException("Phase " + phaseName + " not valid"));
  }
}

File: src/main/kotlin/com/matag/game/turn/phases/PhaseFactory.kt
package com.matag.game.turn.phases

import org.springframework.stereotype.Component

@Component
class PhaseFactory(
    private val phases: List<Phase>
) {

    fun get(phaseName: String): Phase =
        phases.firstOrNull { it.name == phaseName } ?: throw UnsupportedOperationException("Phase $phaseName not valid")
}
